Created Driverless AI Scoring Pipeline Deployment to MiNiFi C++

james94 opened the pull request:
This PR is based on a previous PR, which I closed to address certain issues: [Integrates H2O Driverless AI MOJO & Python Scoring Pipeline into Apache MiNiFi C++ #20](https://github.com/h2oai/dai-deployment-examples/pull/20)

The main Driverless AI Deployment example that is illustrated is **deploying the MOJO Scoring Pipeline
to MiNiFi C++**. The second deployment example if users are interested is deploying the Python Scoring
Pipeline to MiNiFi C++. For this deployment example, I included a **[conf/](https://github.com/james94/dai-deployment-examples/tree/mojo-py-minificpp/mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/conf)** directory that has prebuilt
MiNiFi Data Flows, **[minifi-python/h2o](https://github.com/james94/dai-deployment-examples/tree/mojo-py-minificpp/mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o)** directory that has the custom MiNiFi Python Processors for
Driverless AI Scoring Pipeline Deployment. I also include a **[modules/](https://github.com/james94/dai-deployment-examples/tree/mojo-py-minificpp/mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/modules)** directory with a Python script
that imports the Driverless AI Python Scorer since the filename changes frequently and that way users
can update that file instead of the custom MiNiFi Python Processor. There is **1 custom MOJO MiNiFi Python
Processor**, **2 custom Driverless AI Python Scorer MiNiFi Python Processors** and **1 custom Python Datatable
MiNiFi Python Processor**.

For Driverless AI MOJO Deployment to MiNiFi C++ using MOJO Python Wrapper, visit the [README.md](https://github.com/james94/dai-deployment-examples/tree/mojo-py-minificpp/mojo-py-minificpp)

- [ExecuteDaiMojoScoringPipeline.py](https://github.com/james94/dai-deployment-examples/blob/mojo-py-minificpp/mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/msp/ExecuteDaiMojoScoringPipeline.py)

For Driverless AI Python Scorer Deployment to MiNiFi C++ using Python Runtime, visit the [minifi-python-scoring-pipeline.md](https://github.com/james94/dai-deployment-examples/blob/mojo-py-minificpp/mojo-py-minificpp/minifi-python-scoring-pipeline.md)

- [ExecuteDaiPspBatchScoring.py](https://github.com/james94/dai-deployment-examples/blob/mojo-py-minificpp/mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/psp/ExecuteDaiPspBatchScoring.py)

- [ExecuteDaiPspRealTimeScoring.py](https://github.com/james94/dai-deployment-examples/blob/mojo-py-minificpp/mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/psp/ExecuteDaiPspRealTimeScoring.py)

nkpng2k commented on mojo-py-minificpp/minifi-python-scoring-pipeline.md:
```diff
@@ -0,0 +1,405 @@
+# Execute the Driverless AI Python Scoring Pipeline within MiNiFi Data Flow
```
why 2 readmes?

nkpng2k replied:
too complicated, customer will never get through the first, let alone the second.

nkpng2k replied:
also this looks more or less like a duplicate of the first readme.

james94 replied:
@nkpng2k deploying the DAI Python Scoring Pipeline is slightly different than deploying the DAI MOJO Scoring Pipeline to MiNiFi C++ Data Flow. So, I have 2 READMEs to stay organized. I can work on automating the environment setup for both deployment options. Ideally the customer wouldn't be using MiNiFi C++ directly like I did in the README, they would build their MiNiFi Data Flow using Cloudera Edge Manager (CEM). CEM allows for users to build their MiNiFi Data Flows in a visual canvas similar to NiFi, but for edge device data flows.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/conf/config-psp-interactive-scoring.yml:
```diff
@@ -0,0 +1,76 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
hmmm so I actually dont know about the licensing in `dai-deployment-examples` I would in remove the licensing in the files here, because they may conflict with top level repo licensing... I doubt it though, but good to be careful

james94 replied:
@nkpng2k I removed the licensing in the files. This update will be in the new commit.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/modules/DaiPythonScorer.py:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env python
+
+from scoring_h2oai_experiment_0abac7c4_8993_11ea_ae54_0242ac110002 import Scorer
```
this will not work without that exact mojo python scoring pipeline....

nkpng2k replied:
I'd advocate for a python script generator maybe... or just a readme that tells user how to do this...

james94 replied:
@nkpng2k for now, I removed the .md file that explained how to deploy the DAI Python Scoring Pipeline in MiNiFi C++ Data Flow and also the DaiPythonScorer.py file. I still kept the python processor code for these two DAI Python Scoring Pipeline processors. The focus will be on the MOJO Scoring Pipeline integration with MiNiFi C++.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/psp/ExecuteDaiPspRealTimeScoring.py:
```diff
@@ -0,0 +1,119 @@
+        self.content = codecs.getreader('utf-8')(input_stream).read()
+        return len(self.content)
+
+class ContentWrite(object):
```
qq - required by minifi correct? these class names are just not very descriptive.

james94 replied:
@nkpng2k I verified that the updated class names work great in `ExecuteDaiMojoScoringPipeline.py` code. My priority for the latest updates was focused around the DAI MOJO Scoring Pipeline processor.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/psp/ExecuteDaiPspRealTimeScoring.py:
```diff
@@ -0,0 +1,119 @@
+        pred_header_str = context.getProperty("Predicted Label(s)")
+        pred_header_list = pred_header_str.split(",")
+        # load tabular data str into datatable, convert to pd df, then to list of lists
+        test_dt_frame = dt.Frame(read_cb.content)
```
I would actually advocate that you write a separated class def for say `DaiMojoScorer` and it would include methods for importing data, transforming it, scoring it, etc... 

one massive function is hard to read and change later.

james94 replied:
@nkpng2k thanks for the suggestion. For now I split up parts of this long `onTrigger` function into smaller functions to handle importing data, transforming it, scoring it and transmitting the scores.

nkpng2k commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/psp/ExecuteDaiPspRealTimeScoring.py:
```diff
@@ -0,0 +1,119 @@
+
+scorer = None
+
+def describe(processor):
```
I'd advocate for like a wrapper class to house all these methods... but thats assuming its allowed.

james94 replied:
@nkpng2k I am not sure if it is allowed. I could look into incorporating a wrapper class for a future version of this python processor. For now, I will stay with using multiple functions that make the code easier to follow.

nkpng2k reviewed (dismissed):
I think small changes. Would recommend maybe a requirements.txt if there are any dependencies coming out of the nifi/minifi stuffs that would be useful. 

also maybe a setup.py? idk if minifi allows for installing of whl files.

orendain commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/msp/ExecuteDaiMojoScoringPipeline.py:
```diff
@@ -0,0 +1,130 @@
+import datatable as dt
+from collections import Counter
+from scipy.special._ufuncs import expit
+import daimojo.model
```
Just as a random suggestion: Consider the use of a code formatter.  I've been recommended [Black](https://black.readthedocs.io/en/stable/) by Python engineers I highly respect the opinions of.

james94 replied:
@orendain thanks I will give it a try.

orendain commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/dai/msp/ExecuteDaiMojoScoringPipeline.py:
```diff
@@ -0,0 +1,130 @@
+        # add flow file attribute: number of rows to know how many rows were scored
+        flow_file.addAttribute("num_rows_scored", str(preds_dt_frame.nrows))
+        # add one or more flow file attributes: predicted label name and associated score pair
+        for i in range(len(pred_header)):
+            ff_attr_name = pred_header[i] + "_pred_0"
+            flow_file.addAttribute(ff_attr_name, str(preds_df.at[0,pred_header[i]]))
+            log.info("getAttribute({}): {}".format(ff_attr_name, flow_file.getAttribute(ff_attr_name)))
```
optional:
Instead of using indices, I'd consider iterating over the headers themselves.  That way, you don't have to lookup the strings with `pred_header[i]`.  You would just reference `header` if you do something like `for header in pred_header`.

james94 replied:
@orendain good point. I will make this change and make sure it is in the new commit.

orendain commented on mojo-py-minificpp/README.md:
```diff
@@ -0,0 +1,330 @@
+2\. Set **EC2 Public DNS** and **EC2 Pem Key** as permanent environment variables
+
+~~~bash
+# For Mac OS X, set permanent environment variables 
+tee -a $HOME/.bash_profile << EOF
+# Set EC2 Public DNS
+export H2O_DAI_SCORING_INSTANCE={EC2 Public DNS}.compute.amazon.com
+# Set EC2 Pem Key
+export H2O_DAI_SCORING_PEM=$HOME/.ssh/{private-key-filename}.pem
```
For this and other code blocks, I'd consider removing a lot of the in-code comments.  The readmes are already quite long as it is, and removing comments that are either obvious or already described before the code, may save many lines of reading.

james94 replied:
@orendain for that code block, I removed most of the comments and I also used echo "" | tee -a path/to/.profile since it saved space.

orendain commented on mojo-py-minificpp/README.md:
```diff
@@ -0,0 +1,330 @@
+# Deploy Driverless AI MOJO within a MiNiFi C++ Data Flow
+
+## Cloudera Integration Point for CDF
```
Since the readmes are relatively long, I'd consider having a "Contents" (or similar) section with a heirarchy of the included sections with links to them.  So the reader knows what to expect.

james94 replied:
@orendain in the new commit, the README for deploying DAI MOJO SP within a MiNiFi C++ Data Flow is much shorter now. I created a `setup.sh` to automate almost all of the install.

james94 commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/conf/config-msp-batch-scoring.yml:
```diff
@@ -0,0 +1,76 @@
+# distributed under the License is distributed on an \"AS IS\" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
@orendain I removed all instances of Apache License 2. Those updates will be in the new commit.

james94 commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/conf/minifi.properties:
```diff
@@ -0,0 +1,84 @@
+## very likely the same base url of rest urls
+#nifi.c2.flow.base.url=
+#nifi.c2.rest.url=
+#nifi.c2.rest.url.ack=
```
@orendain thank you for the suggestion. I made the update and it will be in the new commit.

james94 commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/ConvertDsToCsv.py:
```diff
@@ -0,0 +1,86 @@
+
+    pip install datatable
+    pip install pandas
+"""
```
@orendain thanks for the suggestion. I will make sure to have these instructions for dependencies removed from the python script.

james94 commented on mojo-py-minificpp/model-deployment/apps/nifi-minifi-cpp/minifi-python/h2o/ConvertDsToCsv.py:
```diff
@@ -0,0 +1,86 @@
+    """
+    flow_file = session.get()
+    if flow_file is not None:
+        read_cb = ContentExtract()
```
@orendain thank you for the recommendation. I made this update for **ExecuteDaiMojoScoringPipeline.py** processor. I will be removing **ConvertDsToCsv.py** since it is now on the official nifi-minifi-cpp github repo:

- [ConvertDsToCsv.py](https://github.com/apache/nifi-minifi-cpp/blob/main/extensions/pythonprocessors/h2o/ConvertDsToCsv.py)

james94 commented:
@nkpng2k @orendain c7817c9 

This new commit addresses most of your feedback. I narrowed down the README, so users who choose this DAI MOJO SP Deployment to MiNiFi C++ Agent option can follow along much easier. I reduced the amount of commands I was using in the previous README and moved many of them over to a setup.sh script. I have included fewer commands for sending Driverless AI MOJO SP assets to the dev EC2 environment, so then user can run setup.sh to install everything: Cloudera Edge Management (includes Edge Flow Manager, NiFi Registry and one MiNiFi C++ Agent) and DAI MOJO2 Python Runtime on a single node. After the installation, the user is taken to EFM UI for building a data flow that will deploy a DAI MOJO SP and once the data flow is built, they can use EFM to publish the data flow to a MiNiFi C++ Agent running on the same EC2 environment. Alternatively, a user could install MiNiFi C++ Agent on an edge device, such as a Raspberry Pi, Nvidia Jetson Nano, etc and publish a data flow there too for edge intelligence. For simplicity, I keep everything in the same environment.

james94 commented:
@nkpng2k I believe my latest commit addresses many of your recommendations. Can you re-review my PR, so we can get closer to merging this PR? I incorporated a mojo2-py-requirements.txt and setup.sh in this PR. I focused on improving the experience for user deploying DAI MOJO SP in MiNiFi C++ Data Flow. I removed the .md file related to python scoring pipeline, but kept the python scoring pipeline related processors in case anyone is interested. You will also find with the latest MOJO scoring pipeline related processor, a lot of the code that was in the `onTrigger()` function has been moved to other functions to reduce the size of the `onTrigger()` function and make it easier to read and maintain.

james94 commented:
@nkpng2k I just want to give you a heads up that I am going to merge this PR into master. I made sure to add a mojo2-py-requirements.txt, setup.sh and shortened the guide, so it is easier for users to follow along with. I added EFM, which allows users to build MiNiFi C++ data flows in a GUI like NiFi, but for edge devices. Feel free to look at this contribution later when you have time.

james94 commented:
Thank you for your feedback. @nkpng2k @orendain
